File: utils/logger.py
"""
时间记录器后端日志模块
统一处理所有后端日志记录功能
"""

import os
import json
import threading
from datetime import datetime, timedelta

# 日志存储文件路径
LOG_FILE = os.path.join('data', 'logs.txt')
LOG_RETENTION_HOURS = 2  # 保留2小时内的日志

# 使用线程锁确保线程安全
_log_lock = threading.Lock()

class TimeRecorderLogger:
    """时间记录器后端日志类"""
    
    @staticmethod
    def _get_logs():
        """获取所有日志"""
        try:
            if os.path.exists(LOG_FILE):
                with open(LOG_FILE, 'r', encoding='utf-8') as f:
                    # 读取所有日志行
                    lines = f.readlines()
                    logs = []
                    for line in lines:
                        # 解析日志行
                        if line.strip():
                            logs.append(line.strip())
                    return logs
            return []
        except Exception as e:
            print(f"读取日志文件失败: {e}")
            return []
    
    @staticmethod
    def _save_logs(logs):
        """保存日志到文件"""
        try:
            # 确保data目录存在
            os.makedirs(os.path.dirname(LOG_FILE), exist_ok=True)
            
            # 清理过期日志
            logs = TimeRecorderLogger._cleanup_old_logs(logs)
            
            with open(LOG_FILE, 'w', encoding='utf-8') as f:
                for log in logs:
                    f.write(log + '\n')
        except Exception as e:
            print(f"保存日志文件失败: {e}")
    
    @staticmethod
    def _cleanup_old_logs(logs):
        """清理过期日志（仅保留2小时内的日志）"""
        cutoff_time = datetime.now() - timedelta(hours=LOG_RETENTION_HOURS)
        
        # 过滤掉过期的日志
        filtered_logs = []
        for log in logs:
            try:
                # 从日志行中提取时间戳
                # 格式: [2025-10-19T05:35:07.091058] [INFO] [test_module] 验证日志清理功能
                if log.startswith('['):
                    end_timestamp = log.find(']')
                    if end_timestamp > 0:
                        timestamp_str = log[1:end_timestamp]
                        log_time = datetime.fromisoformat(timestamp_str.replace('Z', '+00:00'))
                        if log_time >= cutoff_time:
                            filtered_logs.append(log)
            except Exception:
                # 如果时间戳解析失败，保留该日志
                filtered_logs.append(log)
        
        return filtered_logs
    
    @staticmethod
    def log(level, module, message, data=None):
        """
        记录日志
        :param level: 日志级别 (debug, info, warn, error)
        :param module: 模块名称
        :param message: 日志消息
        :param data: 附加数据（可选）
        """
        # 创建日志条目
        timestamp = datetime.now().isoformat()
        log_entry = f"[{timestamp}] [{level.upper()}] [{module}] {message}"
        
        # 如果有附加数据，添加到日志条目中
        if data is not None:
            try:
                # 尝试将数据转换为JSON格式
                data_str = json.dumps(data, ensure_ascii=False)
                log_entry += f" {data_str}"
            except Exception:
                # 如果不能序列化，转换为字符串
                log_entry += f" {str(data)}"
        
        # 输出到控制台
        console_message = log_entry
        if level == 'error':
            print(f"ERROR: {console_message}")
        elif level == 'warn':
            print(f"WARNING: {console_message}")
        else:
            print(console_message)
        
        # 保存到文件（线程安全）
        with _log_lock:
            logs = TimeRecorderLogger._get_logs()
            logs.append(log_entry)
            TimeRecorderLogger._save_logs(logs)
    
    @staticmethod
    def debug(module, message, data=None):
        """记录调试日志"""
        TimeRecorderLogger.log('debug', module, message, data)
    
    @staticmethod
    def info(module, message, data=None):
        """记录信息日志"""
        TimeRecorderLogger.log('info', module, message, data)
    
    @staticmethod
    def warn(module, message, data=None):
        """记录警告日志"""
        TimeRecorderLogger.log('warn', module, message, data)
    
    @staticmethod
    def error(module, message, data=None):
        """记录错误日志"""
        TimeRecorderLogger.log('error', module, message, data)
    
    @staticmethod
    def get_logs():
        """获取所有日志（已清理过期日志）"""
        with _log_lock:
            logs = TimeRecorderLogger._get_logs()
            return TimeRecorderLogger._cleanup_old_logs(logs)
    
    @staticmethod
    def clear_logs():
        """清空所有日志"""
        with _log_lock:
            try:
                if os.path.exists(LOG_FILE):
                    os.remove(LOG_FILE)
            except Exception as e:
                print(f"清空日志文件失败: {e}")
    
    @staticmethod
    def get_logs_by_module(module):
        """获取指定模块的日志"""
        logs = TimeRecorderLogger.get_logs()
        # 在文本日志中查找包含模块名的日志
        return [log for log in logs if f"[{module}]" in log]
    
    @staticmethod
    def get_logs_by_level(level):
        """获取指定级别的日志"""
        logs = TimeRecorderLogger.get_logs()
        # 在文本日志中查找包含级别名的日志
        return [log for log in logs if f"[{level.upper()}]" in log]

File: static/js/modules/logger.js
/**
 * 时间记录器日志模块
 * 统一处理所有前端日志记录功能
 */

const LOG_STORAGE_KEY = 'timeRecorderLogs';
const LOG_RETENTION_HOURS = 2; // 保留2小时内的日志

// 添加发送日志到后端的函数
function sendLogToBackend(logText) {
    // 创建一个隐藏的表单来发送日志到后端
    const formData = new FormData();
    formData.append('log', logText);
    
    // 发送到后端API
    fetch('/api/logs/frontend', {
        method: 'POST',
        body: formData
    }).catch(err => {
        // 静默处理错误，避免影响用户体验
        console.debug('发送前端日志到后端失败:', err);
    });
}

export const TimeRecorderLogger = {
    /**
     * 记录日志
     * @param {string} level - 日志级别 (debug, info, warn, error)
     * @param {string} module - 模块名称
     * @param {string} message - 日志消息
     * @param {object} data - 附加数据（可选）
     */
    log: function(level, module, message, data = null) {
        const timestamp = new Date().toISOString();
        const logEntry = {
            timestamp: timestamp,
            level: level,
            module: module,
            message: message,
            data: data
        };
        
        // 同时输出到控制台
        const consoleMessage = `[${timestamp}] [${level.toUpperCase()}] [${module}] ${message}`;
        switch (level) {
            case 'debug':
                console.debug(consoleMessage, data || '');
                break;
            case 'info':
                console.info(consoleMessage, data || '');
                break;
            case 'warn':
                console.warn(consoleMessage, data || '');
                break;
            case 'error':
                console.error(consoleMessage, data || '');
                break;
            default:
                console.log(consoleMessage, data || '');
        }
        
        // 格式化日志文本
        let logText = `[${timestamp}] [${level.toUpperCase()}] [${module}] ${message}`;
        
        // 如果有附加数据，添加到日志条目中
        if (data !== null) {
            try {
                // 尝试将数据转换为JSON格式
                const dataStr = JSON.stringify(data);
                logText += ` ${dataStr}`;
            } catch (e) {
                // 如果不能序列化，转换为字符串
                logText += ` ${String(data)}`;
            }
        }
        
        // 保存到本地存储
        this._saveLog(logText);
        
        // 发送到后端
        sendLogToBackend(logText);
    },
    
    /**
     * 记录调试日志
     * @param {string} module - 模块名称
     * @param {string} message - 日志消息
     * @param {object} data - 附加数据（可选）
     */
    debug: function(module, message, data = null) {
        this.log('debug', module, message, data);
    },
    
    /**
     * 记录信息日志
     * @param {string} module - 模块名称
     * @param {string} message - 日志消息
     * @param {object} data - 附加数据（可选）
     */
    info: function(module, message, data = null) {
        this.log('info', module, message, data);
    },
    
    /**
     * 记录警告日志
     * @param {string} module - 模块名称
     * @param {string} message - 日志消息
     * @param {object} data - 附加数据（可选）
     */
    warn: function(module, message, data = null) {
        this.log('warn', module, message, data);
    },
    
    /**
     * 记录错误日志
     * @param {string} module - 模块名称
     * @param {string} message - 日志消息
     * @param {object} data - 附加数据（可选）
     */
    error: function(module, message, data = null) {
        this.log('error', module, message, data);
    },
    
    /**
     * 保存日志到本地存储
     * @param {object} logEntry - 日志条目
     */
    _saveLog: function(logEntry) {
        try {
            // 获取现有日志
            let logs = this._getLogs();
            
            // 添加新日志
            logs.push(logEntry);
            
            // 清理过期日志
            logs = this._cleanupOldLogs(logs);
            
            // 保存到本地存储
            localStorage.setItem(LOG_STORAGE_KEY, JSON.stringify(logs));
            
            // 同时发送到后端保存到文件
            this._sendLogToBackend(logEntry);
        } catch (e) {
            console.error('保存日志失败:', e);
        }
    },
    
    /**
     * 获取所有日志
     * @returns {Array} 日志数组
     */
    _getLogs: function() {
        try {
            const logsStr = localStorage.getItem(LOG_STORAGE_KEY);
            return logsStr ? JSON.parse(logsStr) : [];
        } catch (e) {
            console.error('读取日志失败:', e);
            return [];
        }
    },
    
    /**
     * 清理过期日志（仅保留2小时内的日志）
     * @param {Array} logs - 日志数组
     * @returns {Array} 清理后的日志数组
     */
    _cleanupOldLogs: function(logs) {
        const cutoffTime = new Date(Date.now() - LOG_RETENTION_HOURS * 60 * 60 * 1000);
        
        return logs.filter(log => {
            const logTime = new Date(log.timestamp);
            return logTime >= cutoffTime;
        });
    },
    
    /**
     * 获取所有日志
     * @returns {Array} 日志数组
     */
    getAllLogs: function() {
        return this._getLogs();
    },
    
    /**
     * 获取指定模块的日志
     * @param {string} module - 模块名称
     * @returns {Array} 日志数组
     */
    getLogsByModule: function(module) {
        const logs = this._getLogs();
        return logs.filter(log => log.module === module);
    },
    
    /**
     * 获取指定级别的日志
     * @param {string} level - 日志级别
     * @returns {Array} 日志数组
     */
    getLogsByLevel: function(level) {
        const logs = this._getLogs();
        return logs.filter(log => log.level === level);
    },
    
    /**
     * 清空所有日志
     */
    clearLogs: function() {
        localStorage.removeItem(LOG_STORAGE_KEY);
    },
    
    /**
     * 导出日志为文本格式
     * @returns {string} 日志文本
     */
    exportLogs: function() {
        const logs = this._getLogs();
        return logs.join('\n');
    }
};